Fix accurate remix hydro [ANT-1825]

Juliette-Gerbaux opened the pull request:
# Major changes
- Always enable level tracking for hydro reservoirs in the weekly problems, instead of recomputing the level as a post-processing (function `computingHydroLevels`). In particular, always instantiate the following hourly variables
    - level
    - turbining
    - pumping (if available)
    - overflow
- Add overflow cost, expressed as a difference to the spillage cost (property `overflow spilled cost difference`)
- Add a hard-coded cost for levels `LEVEL_COST = -1.e-6` for the reservoir levels to avoid overflows when a reservoir is not full
- Take overflows into account in the daily heuristic step

# Minor changes
- Minor changes/fixes on accurate shave peaks algorithm: take pumping efficiency into account
- Apart from water values, always use reservoir levels in MWh (0..reservoirCapacity). Remove conversions from/to %. This change shouldn't change the final results, except maybe for rounding errors.
- Add a small tolerance for reservoir level checks (`RESERVOIR_LEVEL_TOLERANCE`) to avoid false positives caused by numerical artifacts

flomnes committed:
Save results for overflow

flomnes committed:
Save results for overflow

flomnes committed:
Disable tests (windows)

flomnes committed:
Try boost fix

flomnes committed:
Disable tests

Commit:
Remove truncation

Commit:
Change overflow cost

Commit:
Change overflow cost

Commit:
Change overflow cost

Commit:
Change overflow cost

flomnes committed:
Fix infeasibility in H2O_J problem

flomnes committed:
Remove HydroLevelsUpdatePostProcessCmd, etc.

flomnes committed:
Add overflow cost

flomnes committed:
SuiviNiveauHoraire = true

flomnes committed:
Remove unused class computeTimeStepLevel

Commit:
Merge remote-tracking branch 'github/develop' into fix/overflow

Commit:
Post-merge

Commit:
Merge remote-tracking branch 'github/fix/overflow' into fix/overflow

Commit:
Merge remote-tracking branch 'github/develop' into fix/overflow

Commit:
Remove forgotten HydroLevelsUpdatePostProcessCmd

Commit:
Fix

Commit:
Cout de remplissage

Commit:
Counting

Commit:
Re-introduce computingHydroLevels, only for scaling

This reverts commit ac97a130974db05179489ccf14f3eefde7948d27.

Commit:
CoutDeRemplissage

flomnes committed:
Fix

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/overflow

flomnes committed:
Disable some tests

flomnes committed:
Add reset for levelCost

flomnes committed:
Add load for levelCost

flomnes committed:
Merge remote-tracking branch 'origin/develop' into HEAD

flomnes committed:
Fix build

Juliette-Gerbaux committed:
Remove truncation

Commit:
Add logs

Commit:
Edit logs

Commit:
Edit logs

Commit:
Change rhs hydro constraint

Commit:
Add efficiency and reservoir management

Commit:
Increase tolerance

Commit:
Time remix hydro

flomnes committed:
Improve naming, fix unit tests, clean up

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/logs-accurate-remix-hydro

flomnes committed:
Clean up logs

flomnes committed:
Fix &&

flomnes committed:
Use std::tie

flomnes committed:
Remove a few HydroLevelsUpdatePostProcessCmd/computingHydroLevels

flomnes committed:
Hard code a parameter, rename another

flomnes committed:
Partial revert

flomnes committed:
Partial revert

flomnes committed:
Enable tests

flomnes committed:
Fix

flomnes committed:
Review

flomnes committed:
Remove scaling of levels

flomnes committed:
Remove useless

flomnes committed:
Remove std::vector containing a constant data

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/logs-accurate-remix-hydro

flomnes committed:
Update migration guide & such

flomnes committed:
use constexpr

flomnes committed:
Update existing tests

JasonMarechal25 commented:
This pull request includes several changes to the hydraulic reservoir management and optimization process. The most important changes involve adding new properties, modifying default values, and removing redundant code.

### Hydraulic Reservoir Management:

* Added a new property `overflow spilled cost difference` to `input/hydro/hydro.ini` and set its default value to 1.0. This property describes the additional cost of overflow relative to the cost of spillage for each area. (`docs/user-guide/04-migration-guides.md`, `src/libs/antares/study/include/antares/study/parts/hydro/container.h`, `src/libs/antares/study/parts/hydro/container.cpp`) [[1]](diffhunk://#diff-4a251a9fc9c1b42858f26e5549c87a0f700f577279cfdaf4f6cabd9ef7499be8R24-R27) [[2]](diffhunk://#diff-11a80860303872aa49561428d57b68b75ee972fb1b7426b2d0d4e4e11f9efa38R232-R233) [[3]](diffhunk://#diff-93076193be523abdd76d790955d60f58b0d77cc2024d1834ae229ae119497b74R72) [[4]](diffhunk://#diff-93076193be523abdd76d790955d60f58b0d77cc2024d1834ae229ae119497b74R328-R336) [[5]](diffhunk://#diff-93076193be523abdd76d790955d60f58b0d77cc2024d1834ae229ae119497b74R509) [[6]](diffhunk://#diff-93076193be523abdd76d790955d60f58b0d77cc2024d1834ae229ae119497b74L515-R527) [[7]](diffhunk://#diff-93076193be523abdd76d790955d60f58b0d77cc2024d1834ae229ae119497b74R584-R585)

### Default Values and Parameters:

* Changed the default value of the `shedding.policy` property in `settings/generaldata.ini` from `shave peaks` to `accurate shave peaks`. (`src/libs/antares/study/parameters.cpp`)

### Code Cleanup and Optimization:

* Removed the `HydroLevelsUpdatePostProcessCmd` class and related calls, as it is no longer needed. (`src/solver/optimisation/include/antares/solver/optimisation/post_process_commands.h`, `src/solver/optimisation/adequacy_patch_csr/adq_patch_post_process_list.cpp`, `src/solver/optimisation/optim_post_process_list.cpp`, `src/solver/optimisation/post_process_commands.cpp`, `src/solver/simulation/adequacy.cpp`) [[1]](diffhunk://#diff-f3aacde720bd688fb00c61453c35eb406d75e02a051f4e958dd6b62a0c0f72d5L40-L54) [[2]](diffhunk://#diff-285dd45e13548a7f7e64425c270c50fae7688aa331a89eb3d7487d6c807e6d71L40-L41) [[3]](diffhunk://#diff-285dd45e13548a7f7e64425c270c50fae7688aa331a89eb3d7487d6c807e6d71L55-L56) [[4]](diffhunk://#diff-93400db61d68471bcb1029573fccac7b03d7f1c051dec4872ce04fca5789cb49L40-L48) [[5]](diffhunk://#diff-5c91254730453e0c6cf6ec40138110210e5a02a2dfef890fe27cfb3f21786031L72-L90) [[6]](diffhunk://#diff-d138318f932367ffbc430e853fcabff573b0988fbfa0f435844a2a13082bf60dL216-L225)

### Cost and Constraints Management:

* Introduced `LEVEL_COST` to avoid indetermination of the overflow variable, ensuring overflows occur as late as possible. (`src/solver/optimisation/opt_gestion_des_couts_cas_lineaire.cpp`) [[1]](diffhunk://#diff-8f1e81dd6e5322a8a9601c7ead8971e72bbb4445b2fa5faca49baa9f8a9ce5e5R26-R27) [[2]](diffhunk://#diff-8f1e81dd6e5322a8a9601c7ead8971e72bbb4445b2fa5faca49baa9f8a9ce5e5R297-R313)
* Simplified the constraints counting process by removing the `MonitorHourlyLev` checks. (`src/solver/optimisation/opt_decompte_variables_et_contraintes.cpp`) [[1]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL133-R142) [[2]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL155-R162) [[3]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL171-R171) [[4]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL180-R180) [[5]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL190-R190) [[6]](diffhunk://#diff-eb42ab93751d1703fb753e7d195869d42d7239b5b04b1289362905d9faf27aadL206-L211)

### Miscellaneous:

* Removed unused variable `TurbineCible` from `H2O_J_InitialiserLesBornesdesVariables`. (`src/solver/hydro/daily/h2o_j_initialiser_les_bornes_des_variables.cpp`) [[1]](diffhunk://#diff-890144218c25ac9bf6b1c3a063a04382dbb8ee1fab9e2c202bc897265ab024fbL32) [[2]](diffhunk://#diff-890144218c25ac9bf6b1c3a063a04382dbb8ee1fab9e2c202bc897265ab024fbL56-R55)
* Updated the `prepareDailyOptimalGenerations` method to include reservoir capacity and problem overflows. (`src/solver/hydro/management/daily.cpp`)
* Adjusted the initialization of overflow variables to store optimized values. (`src/solver/optimisation/opt_gestion_des_bornes_cas_lineaire.cpp`)

JasonMarechal25 commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -345,7 +345,7 @@ void Parameters::reset()
     // Shedding strategies
     power.fluctuations = lssFreeModulations;
-    shedding.policy = shpShavePeaks;
+    shedding.policy = shpAccurateShavePeaks;
```
@flomnes 
Are there some precautions to take when changing the default value of a parameter like this ?

flomnes replied:
It will change the results. But the whole point of this PR is to change the results.

More seriously, the "shave peaks" algorithm generates non-conforming results, so we'd best avoid using it (deprecated property).

flomnes replied:
done

JasonMarechal25 commented on src/solver/optimisation/opt_decompte_variables_et_contraintes.cpp:
```diff
@@ -130,16 +130,16 @@ int OPT_DecompteDesVariablesEtDesContraintesDuProblemeAOptimiser(PROBLEME_HEBDO*
         }
 
-        if (Pump && !TurbEntreBornes && !MonitorHourlyLev)
+        ProblemeAResoudre->NombreDeContraintes += nombreDePasDeTempsPourUneOptimisation;
```
This needs to be explained

flomnes replied:
I added this comment
```
// This estimate is not very accurate, some constraints may not be enabled
// in some cases.
// The counting below should be removed, and we should instead use other methods
// involving dynamic reallocations, etc.
```

JasonMarechal25 commented on src/solver/optimisation/opt_gestion_des_couts_cas_lineaire.cpp:
```diff
@@ -23,6 +23,8 @@
 
 #include "variables/VariableManagerUtils.h"
 
+constexpr double LEVEL_COST = -1.e-6;
```
```suggestion
namespace {
    constexpr double LEVEL_COST = -1.e-6;
}
```

JasonMarechal25 commented on src/solver/simulation/shave-peaks-by-remix-hydro.cpp:
```diff
@@ -133,10 +137,14 @@ static void checkInputCorrectness(const std::vector<double>& DispatchGen,
     {
-        throw std::invalid_argument(msg_prefix
-                                    + "levels computed from input don't respect reservoir bounds");
+        if (!(levels <= reservoirCapacity + RESERVOIR_LEVEL_TOLERANCE)
+            || !(levels >= -RESERVOIR_LEVEL_TOLERANCE))
```
```suggestion
    if (levels > reservoirCapacity + RESERVOIR_LEVEL_TOLERANCE ||
        levels < -RESERVOIR_LEVEL_TOLERANCE)
```

flomnes replied:
The first condition (> or <) requires all elements to be on one side of the threshold, whereas the second condition (!<= or !>=) only requires at least one element to be outside the range.
-    The first condition (levels > X || levels < Y) is stricter.
-    The second condition (!(levels <= X) || !(levels >= Y)) is more permissive.

Hence what you suggest breaks the tests (no exception thrown when previously one was thrown)

JasonMarechal25 commented on src/solver/simulation/shave-peaks-by-remix-hydro.cpp:
```diff
@@ -235,24 +248,33 @@ RemixHydroOutput shavePeaksByRemixingHydro(const std::vector<double>& DispatchGe
-                    max_pic = capa;
-                    max_creux = *std::min_element(intermediate_level.begin(),
-                                                  intermediate_level.end());
+                    std::vector<double> intermediate_level(levels.begin()
+                                                             + std::min(hourBottom, hourPeak),
+                                                           levels.begin()
+                                                             + std::max(hourBottom, hourPeak));
```
```suggestion
                    auto intermediate_level = levels | std::views::drop(std::min(hourBottom, hourPeak)) 
                                  | std::views::take(std::abs(hourBottom - hourPeak));
```

Drop [begin, min(hourBottom, hour peak] then take abs(hourPeak - hourBottom);
If hourBottom < hourPeak then take hourPeak - hourBottom element starting hourBottom
If hourPeak < hourBottom them take hourBottom - hourPeak element starting hourPeak

flomnes replied:
I find the original code easier to read, and chaining filters isn't supported by all compilers (e.g gcc 10) as @payetvin noticed

JasonMarechal25 replied:
Works fine with gcc10 (in this case) https://godbolt.org/z/4aneqYeKP

I personally find it clearer that pointer arithmetic tbh

JasonMarechal25 replied:
Another point. I don't know the size of intermediate_level and how much time we pass in the function, but using vector(it, it) creates a vector with copy of the values, whereas using filter views operates on the collection.
Performance-wise it may be more interesting to use views

flomnes replied:
I used `std::span`, which avoids copies and is readable

JasonMarechal25 commented on src/solver/simulation/shave-peaks-by-remix-hydro.cpp:
```diff
@@ -235,24 +248,33 @@ RemixHydroOutput shavePeaksByRemixingHydro(const std::vector<double>& DispatchGe
+                    if (hourBottom < hourPeak)
+                    {
+                        max_pic = reservoirCapacity;
+                        max_creux = *std::min_element(intermediate_level.begin(),
+                                                      intermediate_level.end());
```
```suggestion
                        max_creux = *std::ranges::min_element(intermediate_level);
```

JasonMarechal25 commented on src/solver/simulation/shave-peaks-by-remix-hydro.cpp:
```diff
@@ -235,24 +248,33 @@ RemixHydroOutput shavePeaksByRemixingHydro(const std::vector<double>& DispatchGe
+                    {
+                        max_pic = reservoirCapacity
+                                  - *std::max_element(intermediate_level.begin(),
+                                                      intermediate_level.end());
```
```suggestion
                                  - *std::ranges::max_element(intermediate_level);
```

JasonMarechal25 commented:
Maybe update changelog too with relevant information for this PR

flomnes committed:
Review, minor improvements

flomnes committed:
changelog

JasonMarechal25 committed:
Merge branch 'develop' into fix/logs-accurate-remix-hydro

flomnes committed:
Use std::span

JasonMarechal25 committed:
Merge branch 'develop' into fix/logs-accurate-remix-hydro

flomnes committed:
Increase coverage on hydro common data (hydro.ini) [ANT-1825] (#2634)

# Description
- Split large loading function (time-series / INI file)
- Add coverage on new property `overflow spilled cost difference`

# Summary
```cpp
void CommonFixture::writeValidFile()
{
    std::ofstream outfile(hydroIni);
    outfile <<
      R"([overflow spilled cost difference]
east = 1.00000
west = 2.31000
[reservoir]
east = true)";
}

BOOST_FIXTURE_TEST_CASE(test_read_valid_file, CommonFixture)
{
    writeValidFile();
    BOOST_CHECK(load());
    BOOST_CHECK_EQUAL(east->hydro.overflowSpilledCostDifference, 1.0);
    BOOST_CHECK_EQUAL(west->hydro.overflowSpilledCostDifference, 2.31000);
}
```

flomnes committed:
H. LEV in %

flomnes committed:
Level - Capacity

flomnes committed:
Conversion

JasonMarechal25 committed:
Merge branch 'develop' into fix/logs-accurate-remix-hydro
